### PR TITLE
[stdlib] Collapse sequence and collection wrappers

### PR DESCRIPTION
**Pull request by airspeedswift**
Builds on top of #20175 and collapses wrappers like `LazyMapSequence` and `LazyMapCollection` into one now this is possible.

### CODE REVIEW

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test source compatibility

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please smoke test macOS platform

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

**Comment by airspeedswift** [comment]
@swift-ci please benchmark

**Comment by airspeedswift** [comment]
@swift-ci please test

File: stdlib/public/core/Sequence.swift [hunk 1]
```diff
@@ -1124,18 +922,21 @@ extension Sequence where SubSequence == AnySequence<Element> {
...
 
     if i != ringBuffer.startIndex {
-      let s0 = ringBuffer[i..<ringBuffer.endIndex]
-      let s1 = ringBuffer[0..<i]
-      return AnySequence([s0, s1].joined())
+      var rotated: [Element] = []
+      rotated.reserveCapacity(ringBuffer.count)
+      rotated += ringBuffer[i..<ringBuffer.endIndex]
+      rotated += ringBuffer[0..<i]
+      return rotated
```
Comment (palimondo) [inline stdlib/public/core/Sequence.swift]: How does this copying approach compare to in-place rotation?
````swift
ringBuffer[0..<i].reverse()
ringBuffer[i..<ringBuffer.endIndex].reverse()
ringBuffer[0..<ringBuffer.endIndex].reverse()
````

Comment (palimondo) [reply stdlib/public/core/Sequence.swift]: Also, is there a potential perf. win in using `ContiguousArray`s here (and in `dropLast`) and then `return Array(ringBuffer)`? I've noticed that's what `map` and `filter` implementations use.

Comment (airspeedswift) [reply stdlib/public/core/Sequence.swift]: Either are worth exploring to see – they're not important for the ABI though so we can look at trying them out after this change lands.

**Pull request merged**